Improved Battery resistance estimation

rmackay9 opened the pull request:
This PR does the following:

- moves the battery resistance estimation from the copter AP_Motors library into the Battery monitor library so it can be used by all vehicles
- improves the resistance estimation by looking for short-term changes in current and voltage.  The magic behind this is all in the AP_BattMonitor_Backend::update_resistance_estimate.  @lthall is the designer so please ask your questions about how this works to him!
- optionally allows the battery failsafe to use the estimated resting voltage (i.e. voltage with sag caused by current draw removed)
- Dataflash logging of Current adds new resting voltage and resistance fields.
- Dataflash Current logging is split into two messages so that individual cell voltages appear in a new message called "CUC" and "CUC2" (CUrrent Cell).  This was done primarily because the CURR message column names were getting too long.  It also reduces the amount of logging for most users slightly.

Here's is a graph from Leonard showing how the new resistance learning performs better than the older method.
![batt_resistance_estimate](https://cloud.githubusercontent.com/assets/1498098/26615905/7d1378e0-4605-11e7-9dc5-8ed62fee699f.png)

Notice how in the top graph, the new resting voltage estimate (light-green line) is much smoother than with the older method (dark blue line).  Also notice in the bottom graph that the new resistance estimate (dark blue) is also much smoother than the older estimate (dark red).

I (Randy) found during testing on my IRIS that the resistance estimate early on in the flight was far too high leading to unrealistically high resting voltage estimates.  Leonard says he has a fix for this so I've marked this PR as "WIP".
![iris-battvoltandresistance](https://cloud.githubusercontent.com/assets/1498098/26615996/4bf47b3c-4606-11e7-81b7-a3f0072ca4bd.png)

One question is whether we should store the resistance values to a parameter to avoid learning at startup.

lvale commented:
@rmackay9 The resistance is battery dependent. On normal batteries you don't have a way to say which battery has x resistance. On smart batteries, if we can ensure that the battery has a unique identified that would be feasible, but then that value makes sense to be stored on the battery itself, not on the FC.

amilcarlucas commented:
lvale +1

lthall commented:
So this problem is caused by a poorly specified zero current offset on the current monitor so when the current is actually 1A it is only reading 0.5A. This results in relatively large voltage drop for a very small current increase resulting in a very high resistance prediction. You can see from the log that it does learn the resistance once the current becomes larger and the small zero offset is no longer significant.

rmackay9 commented:
One question I have perhaps to @OXINARF and @WickedShell is whether some of these new variables that are only used in the resistance learning and are never exposed to the vehicle code (like _state.voltage_filt, current_max_amps, current_filt_amps and resistance_timer_ms.) should remain in the front-end's BattMonitor_State structure or whether we should instead hide them in the backend.  My feeling is we should move them to the backend... all opinions welcome.

rmackay9 commented:
Fix from Leonard included and tested on the bench.  The graph below shows the improvement.  If you look at the green line (resting voltage) you'll see it never climbs above 13V while in the graph in the image above it climbed to 22V.
![iris-battvoltandresistance-fixed](https://cloud.githubusercontent.com/assets/1498098/26631589/be292028-4646-11e7-9c76-e8f8150d9bc1.png)

I'll do another flight test tomorrow but I suspect the results will be good.

rmackay9 commented:
I've moved those resistance variables into the backend, rebased and re-tested.  The graph below shows that it's doing ok.  It still tends to estimate the resistance as too high during startup but converges on the correct value within about 30seconds.
![battresistanceestimate](https://cloud.githubusercontent.com/assets/1498098/26659804/98cdb312-46ae-11e7-9cbc-92661e64e117.png)

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -356,6 +383,15 @@ bool AP_BattMonitor::overpower_detected(uint8_t instance) const
+bool AP_BattMonitor::has_cell_voltages(const uint8_t instance) const
+{
+    if (instance < _num_instances && drivers[instance] != nullptr) {
+        return drivers[instance]->has_cell_voltages();
```
This is insufficient. It's possible to have a SMBus battery that does't use the fields we are requesting for cell voltages (or doesn't implement them). You need to check that if the driver is capable of fetching a per cell voltage that the cell voltages in the state have actually been set.

OXINARF replied:
The comment from @WickedShell is actually in the wrong place. This method is fine, the one in SMBus backend might not be - I think this is just the equivalent of `has_current` which doesn't say that the current has been read, but that it supports reading current. So I think it is OK that the SMBus backend doesn't check if the cell voltages have actually been read.

One change I think should be done (for `has_current` and `has_cell_voltages`) is to put the methods in each of the SMBus drivers, because logically makes more sense each driver saying what it supports - the base SMBus class provides common methods for reading stuff but it doesn't actually read anything, the drivers do.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -18,6 +18,13 @@
+#define AP_BATT_MONITOR_RES_EST_TC_2        0.1f
+
+// low voltage sources
+#define AP_BATT_MONITOR_LOW_VOLTAGE_SOURCE_RAW              0
```
Please just make these an enum rather then a #define

rmackay9 replied:
sure, changed to enum in follow-up commit.
By the way, I always worry what happens if we cast a parameter value to an enum and it's outside the valid range.  I guess I should test what happens.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp:
```diff
@@ -44,3 +44,54 @@ int32_t AP_BattMonitor_Backend::get_capacity() const
+void AP_BattMonitor_Backend::update_resistance_estimate()
+{
+    // return immediately if no current
+    if (!has_current() || _state.current_amps <= 0.0f) {
```
I'd recommend ```!is_positive(_state.current_amps)``` instead of <= 0.0f. This would allow you to remove the ```is_zero(around _current_max_amps)``` check 7 lines from now.

rmackay9 replied:
great txs.  change to !is_positive in a follow-up commit.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp:
```diff
@@ -44,3 +44,54 @@ int32_t AP_BattMonitor_Backend::get_capacity() const
+    // update maximum current seen since startup and protect against divide by zero
+    _current_max_amps = MAX(_current_max_amps, _state.current_amps);
+    float current_delta = _state.current_amps - _current_filt_amps;
+    if (is_zero(_current_max_amps) || is_zero(current_delta)) {
```
If you change the initial entry function slightly as suggested then ```_current_max_amps``` must be > 0 by the time you get here, and you can discard the check for it.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp:
```diff
@@ -44,3 +44,54 @@ int32_t AP_BattMonitor_Backend::get_capacity() const
+    float filt_alpha = constrain_float(loop_interval/(loop_interval + AP_BATT_MONITOR_RES_EST_TC_1), 0.0f, 0.5f);
+    float resistance_alpha = MIN(1, AP_BATT_MONITOR_RES_EST_TC_2*fabsf((_state.current_amps-_current_filt_amps)/_current_max_amps));
+    float resistance_estimate = -(_state.voltage-_voltage_filt)/current_delta;
+    if (resistance_estimate > 0){
```
```if (is_positive(resistance_estimate))``` would be better, you want to avoid tiny floats that would otherwise fail the is_zero check.

WickedShell commented on libraries/DataFlash/LogStructure.h:
```diff
@@ -920,6 +931,10 @@ Format characters in the format string for binary log messages
     { LOG_CURRENT2_MSG, sizeof(log_Current), \
       "CUR2", CURR_FMT,CURR_LABELS }, \
+    { LOG_CURRENT_CELLS_MSG, sizeof(log_Current_Cells), \
+      "CUCE", CURR_CELL_FMT, CURR_CELL_LABELS }, \
```
I find CUCx to be a really confusing label... I'm not sure what the best one would be but BATx (or BATV before numbering?) might be a bit more intuitive when reviewing logs. (Arguably thats what the CURR one should be, but we can't change it without confusing everyone).

OXINARF replied:
What about CCUR and CCU2? From Cell Current. I would also not say that we can't change the CURR/CUR2 message ID, we have changed IDs in the past - making them BAT/BAT2 would make perfect sense.

rmackay9 replied:
Ok, so will rename CURR, CUR2 to BAT, BAT2.  CUCE, CUC2 will be BCL, BCL2.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -143,15 +143,22 @@ const AP_Param::GroupInfo AP_BattMonitor::var_info[] = {
+    // @Description: Voltage type used for detection of low voltage event
+    // @Values: 0:Raw Voltage, 1:Sag Compensated Voltage
+    // @User: Advanced
+    AP_GROUPINFO("_LOW_TYPE", 22, AP_BattMonitor, _low_voltage_source, 0),
```
You have #define's used at the moment to do the switch later, and while I'd rather see them as an enumerated type this should be using one of the two instead.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -265,6 +269,17 @@ float AP_BattMonitor::voltage(uint8_t instance) const
     }
 }
 
+/// get voltage with sag removed (based on battery current draw and resistance)
```
This description should note that the current pack voltage is returned rather then the compensated one if the user has selected it. The current description implies to me that it can only fetch the compensated voltage, which isn't correct.

rmackay9 replied:
Yes, it's like @OXINARF says.  It's a check that the compensated voltage is reasonable.  I'll modify the comments in the .h and .cpp file to make it clear that "this will always be greater than or equal to the raw voltage"

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor_Analog.h:
```diff
@@ -87,6 +87,9 @@ class AP_BattMonitor_Analog : public AP_BattMonitor_Backend
     /// Read the battery voltage and current.  Should be called at 10hz
     void read();
 
+    /// all smart batteries provide current info
```
Smart batteries in Analog? :slightly_smiling_face:

rmackay9 replied:
fixed, thanks

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor_Backend.h:
```diff
@@ -34,13 +34,30 @@ class AP_BattMonitor_Backend
     AP_BattMonitor                      &_mon;      // reference to front-end
     AP_BattMonitor::BattMonitor_State   &_state;    // reference to this instances state (held in the front-end)
+
+    // resistance estimate
```
Not very important but all of below can be made private, preventing them being changed from backends.

rmackay9 replied:
fixed, thanks!

OXINARF commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -107,10 +107,15 @@ class AP_Motors {
 
+    // get and set battery resistance estimate
+    float               get_batt_resistance() const { return _batt_resistance; }
+    void                set_resistance(float resistance){ _batt_resistance = resistance; }
```
It looks like the only reason these exist is to provide them for logging in MOTB (MotBat), but that can be obtained directly from battery monitor or even removed from this log message since it is already in the battery log messages.

rmackay9 replied:
the motors lib actually uses the resistance in the get_current_limit_max_throttle function.  I definitely agree we've got some duplicate logging of the resistance but I think for support purposes, @lthall would like the resistance to be logged at the same time as the other values used to limit the motor output...

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -128,13 +141,19 @@ class AP_BattMonitor
 
+    // get battery resistance estimate in ohms
+    float get_resistance() const { return get_resistance(AP_BATT_PRIMARY_INSTANCE); }
+    float get_resistance(uint8_t instance) const { return state[instance].resistance; }
```
Nitpicking: to be more correct and match other methods make _instance_  method parameter _const_

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -93,6 +102,10 @@ class AP_BattMonitor
     float voltage() const { return voltage(AP_BATT_PRIMARY_INSTANCE); }
 
+    /// get voltage with sag removed (based on battery current draw and resistance)
+    float voltage_resting_estimate(uint8_t instance) const;
```
Nitpicking: to be more correct and match other methods make _instance_  method parameter _const_

rmackay9 replied:
re consting-ing the instance... but it's pass-by-value because it's a uint8.. so no need to const the argument right?

rmackay9 commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -243,12 +251,8 @@ bool AP_BattMonitor::is_powering_off(uint8_t instance) const {
-                _monitoring[instance] == BattMonitor_TYPE_SOLO ||
-                _monitoring[instance] == BattMonitor_TYPE_BEBOP ||
-                _monitoring[instance] == BattMonitor_TYPE_MAXELL);
+    if (instance < _num_instances && drivers[instance] != nullptr && _monitoring[instance] != BattMonitor_TYPE_NONE) {
```
We check the check of the TYPE quite a bit in this driver and others to allow the driver to be turned off.  So for example the read() call has the same check.

WickedShell replied:
Battery driver changes/disabling requires power cycling realistically. SMBus drivers mostly ignore the read function are still updating all the other values (which in turn propagates out to the rest of the system)

rmackay9 commented on libraries/DataFlash/LogFile.cpp:
```diff
@@ -1924,21 +1924,33 @@ void DataFlash_Class::Log_Write_Current(const AP_BattMonitor &battery)
-        static_assert(ARRAY_SIZE(cells.cells) == (sizeof(pkt.cell_voltages) / sizeof(pkt.cell_voltages[0])),
+        struct log_Current_Cells cell_pkt;
+        const AP_BattMonitor::cells &cells = battery.get_cell_voltages(0);
+        static_assert(ARRAY_SIZE(cells.cells) == (sizeof(cell_pkt.cell_voltages) / sizeof(cell_pkt.cell_voltages[0])),
```
I wasn't sure if the static_assert applied at compile time or runtime.. others tell me it's compile time so I will move it as suggested.

rmackay9 commented:
reworked a bit and re-based on master.

I think I've addressed all comments except:

- the request to move the has-cell-voltages to the _Solo and _Maxell driver level.  I didn't do this because it wouldn't make any functional change and would leave us with dupilcate code.. and I was feeling slightly lazy.
- the request to have has-cell-voltages only return true if the driver has actually received cell level voltages from the battery.  this did cross my mind while developing it and I don't feel strongly about it but this is consistent with the current behaviour and as far as I know, all smart batteries provide cell level voltages. 

Also changed CURR/CUR2 to BAT/BAT2

WickedShell commented:
Fine, I'll send the PR for has-cell-voltages for the backends, because I do feel strongly about it. It's absolutely not a guaranteed response for any generic SMBus battery. :)

Pedals2Paddles commented:
What is the practical use case for the resistance and voltage not under load estimation? Is it so you can make battery alarms and failsafes work of what the battery would be at rest?  This is fascinating and I'm looking forward to testing it out on the Solo
